File: frontend/apps/ios/Pantopus/Core/Networking/Models/Scheduling/ResourceDTOs.swift
//
//  ResourceDTOs.swift
//  Pantopus
//
//  DTOs for bookable home resources — routes `/api/homes/:homeId/scheduling/
//  resources*` (home-only). Resource bookings are render-only via the calendar
//  union (never create HomeCalendarEvent rows). See
//  `reference/calendarly-backend-api.md`.
//

import Foundation

/// A bookable home resource (room / vehicle / tool / charger / other).
public struct ResourceDTO: Decodable, Sendable, Hashable, Identifiable {
    public let id: String
    public let homeId: String?
    public let name: String
    /// `room` | `vehicle` | `tool` | `charger` | `other`.
    public let resourceType: String?
    public let photoURL: String?
    /// `members` | `specific` | `guests`.
    public let whoCanBook: String?
    public let maxDurationMin: Int?
    public let bufferMin: Int?
    public let requiresApproval: Bool?
    public let availableHours: JSONValue?
    public let isActive: Bool?
    public let createdAt: String?
    public let createdBy: String?

    enum CodingKeys: String, CodingKey {
        case id
        case homeId = "home_id"
        case name
        case resourceType = "resource_type"
        case photoURL = "photo_url"
        case whoCanBook = "who_can_book"
        case maxDurationMin = "max_duration_min"
        case bufferMin = "buffer_min"
        case requiresApproval = "requires_approval"
        case availableHours = "available_hours"
        case isActive = "is_active"
        case createdAt = "created_at"
        case createdBy = "created_by"
    }
}

/// A resource booking row (`POST /resources/:rid/book`).
public struct ResourceBookingDTO: Decodable, Sendable, Hashable, Identifiable {
    public let id: String
    public let resourceId: String?
    public let startAt: String?
    public let endAt: String?
    public let name: String?
    public let bookedBy: String?
    public let status: String?
    public let createdAt: String?

    enum CodingKeys: String, CodingKey {
        case id
        case resourceId = "resource_id"
        case startAt = "start_at"
        case endAt = "end_at"
        case name
        case bookedBy = "booked_by"
        case status
        case createdAt = "created_at"
    }
}

/// `GET /resources` → `{ resources }`.
public struct ResourcesResponse: Decodable, Sendable, Hashable {
    public let resources: [ResourceDTO]
}

/// `POST /resources`, `PUT /resources/:rid` → `{ resource }`.
public struct ResourceResponse: Decodable, Sendable, Hashable {
    public let resource: ResourceDTO
}

/// `POST /resources/:rid/book` → 201 `{ booking }`.
public struct ResourceBookingResponse: Decodable, Sendable, Hashable {
    public let booking: ResourceBookingDTO
}

/// Body for `POST /resources`. Owner fields spliced in by the builder.
public struct CreateResourceRequest: Encodable, Sendable {
    public let name: String
    public var resourceType: String?
    public var photoURL: String?
    public var whoCanBook: String?
    public var maxDurationMin: Int?
    public var bufferMin: Int?
    public var requiresApproval: Bool?
    public var availableHours: JSONValue?

    enum CodingKeys: String, CodingKey {
        case name
        case resourceType = "resource_type"
        case photoURL = "photo_url"
        case whoCanBook = "who_can_book"
        case maxDurationMin = "max_duration_min"
        case bufferMin = "buffer_min"
        case requiresApproval = "requires_approval"
        case availableHours = "available_hours"
    }

    public init(
        name: String,
        resourceType: String? = nil,
        photoURL: String? = nil,
        whoCanBook: String? = nil,
        maxDurationMin: Int? = nil,
        bufferMin: Int? = nil,
        requiresApproval: Bool? = nil,
        availableHours: JSONValue? = nil
    ) {
        self.name = name
        self.resourceType = resourceType
        self.photoURL = photoURL
        self.whoCanBook = whoCanBook
        self.maxDurationMin = maxDurationMin
        self.bufferMin = bufferMin
        self.requiresApproval = requiresApproval
        self.availableHours = availableHours
    }
}

/// Body for `PUT /resources/:rid` — partial; at least one field required.
public struct UpdateResourceRequest: Encodable, Sendable {
    public var name: String?
    public var resourceType: String?
    public var photoURL: String?
    public var whoCanBook: String?
    public var maxDurationMin: Int?
    public var bufferMin: Int?
    public var requiresApproval: Bool?
    public var availableHours: JSONValue?

    enum CodingKeys: String, CodingKey {
        case name
        case resourceType = "resource_type"
        case photoURL = "photo_url"
        case whoCanBook = "who_can_book"
        case maxDurationMin = "max_duration_min"
        case bufferMin = "buffer_min"
        case requiresApproval = "requires_approval"
        case availableHours = "available_hours"
    }

    public init(
        name: String? = nil,
        resourceType: String? = nil,
        photoURL: String? = nil,
        whoCanBook: String? = nil,
        maxDurationMin: Int? = nil,
        bufferMin: Int? = nil,
        requiresApproval: Bool? = nil,
        availableHours: JSONValue? = nil
    ) {
        self.name = name
        self.resourceType = resourceType
        self.photoURL = photoURL
        self.whoCanBook = whoCanBook
        self.maxDurationMin = maxDurationMin
        self.bufferMin = bufferMin
        self.requiresApproval = requiresApproval
        self.availableHours = availableHours
    }

    public func encode(to encoder: any Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encodeIfPresent(name, forKey: .name)
        try c.encodeIfPresent(resourceType, forKey: .resourceType)
        try c.encodeIfPresent(photoURL, forKey: .photoURL)
        try c.encodeIfPresent(whoCanBook, forKey: .whoCanBook)
        try c.encodeIfPresent(maxDurationMin, forKey: .maxDurationMin)
        try c.encodeIfPresent(bufferMin, forKey: .bufferMin)
        try c.encodeIfPresent(requiresApproval, forKey: .requiresApproval)
        try c.encodeIfPresent(availableHours, forKey: .availableHours)
    }
}

/// Body for `POST /resources/:rid/book`. Owner fields spliced in by the builder.
public struct BookResourceRequest: Encodable, Sendable {
    public let startAt: String
    public var durationMin: Int?
    public var name: String?

    enum CodingKeys: String, CodingKey {
        case startAt = "start_at"
        case durationMin = "duration_min"
        case name
    }

    public init(startAt: String, durationMin: Int? = nil, name: String? = nil) {
        self.startAt = startAt
        self.durationMin = durationMin
        self.name = name
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/scheduling/ResourceDtos.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.data.api.models.scheduling

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

/**
 * Bookable home resources (home-only). `GET/POST /resources`,
 * `PUT/DELETE /resources/:rid`, `POST /resources/:rid/book`.
 * `resource_type` ∈ `room|vehicle|tool|charger|other`;
 * `who_can_book` ∈ `members|specific|guests` (members only in v1).
 * `available_hours` is a flexible object.
 */
@JsonClass(generateAdapter = true)
data class ResourceDto(
    val id: String,
    @Json(name = "home_id") val homeId: String? = null,
    val name: String,
    @Json(name = "resource_type") val resourceType: String? = null,
    @Json(name = "photo_url") val photoUrl: String? = null,
    @Json(name = "who_can_book") val whoCanBook: String? = null,
    @Json(name = "max_duration_min") val maxDurationMin: Int? = null,
    @Json(name = "buffer_min") val bufferMin: Int? = null,
    @Json(name = "requires_approval") val requiresApproval: Boolean? = null,
    @Json(name = "available_hours") val availableHours: Map<String, Any?>? = null,
    @Json(name = "is_active") val isActive: Boolean? = null,
    @Json(name = "created_at") val createdAt: String? = null,
    @Json(name = "created_by") val createdBy: String? = null,
)

/** `GET /resources` — `{ resources: [...] }`. */
@JsonClass(generateAdapter = true)
data class GetResourcesResponse(
    val resources: List<ResourceDto> = emptyList(),
)

/** `POST/PUT /resources` — `{ resource: … }`. */
@JsonClass(generateAdapter = true)
data class ResourceResponse(
    val resource: ResourceDto,
)

/** Body for `POST /resources`. */
@JsonClass(generateAdapter = true)
data class CreateResourceRequest(
    val name: String,
    @Json(name = "resource_type") val resourceType: String? = null,
    @Json(name = "photo_url") val photoUrl: String? = null,
    @Json(name = "who_can_book") val whoCanBook: String? = null,
    @Json(name = "max_duration_min") val maxDurationMin: Int? = null,
    @Json(name = "buffer_min") val bufferMin: Int? = null,
    @Json(name = "requires_approval") val requiresApproval: Boolean? = null,
    @Json(name = "available_hours") val availableHours: Map<String, Any?>? = null,
    @Json(name = "owner_type") val ownerType: String? = null,
    @Json(name = "owner_id") val ownerId: String? = null,
)

/** Body for `PUT /resources/:rid` (partial). */
@JsonClass(generateAdapter = true)
data class UpdateResourceRequest(
    val name: String? = null,
    @Json(name = "resource_type") val resourceType: String? = null,
    @Json(name = "photo_url") val photoUrl: String? = null,
    @Json(name = "who_can_book") val whoCanBook: String? = null,
    @Json(name = "max_duration_min") val maxDurationMin: Int? = null,
    @Json(name = "buffer_min") val bufferMin: Int? = null,
    @Json(name = "requires_approval") val requiresApproval: Boolean? = null,
    @Json(name = "available_hours") val availableHours: Map<String, Any?>? = null,
)

/** A resource booking (Booking row with `event_type_id=null`). */
@JsonClass(generateAdapter = true)
data class ResourceBookingDto(
    val id: String,
    @Json(name = "resource_id") val resourceId: String? = null,
    @Json(name = "start_at") val startAt: String? = null,
    @Json(name = "end_at") val endAt: String? = null,
    @Json(name = "invitee_name") val inviteeName: String? = null,
    @Json(name = "created_by") val createdBy: String? = null,
    val status: String? = null,
    @Json(name = "created_at") val createdAt: String? = null,
)

/** Body for `POST /resources/:rid/book`. */
@JsonClass(generateAdapter = true)
data class BookResourceRequest(
    @Json(name = "start_at") val startAt: String,
    @Json(name = "duration_min") val durationMin: Int? = null,
    val name: String? = null,
    @Json(name = "owner_type") val ownerType: String? = null,
    @Json(name = "owner_id") val ownerId: String? = null,
)

/** `POST /resources/:rid/book` — `{ booking: … }`. */
@JsonClass(generateAdapter = true)
data class BookResourceResponse(
    val booking: ResourceBookingDto,
)
